src/openforms/payments/contrib/worldline/plugin.py

```diff
@@ -0,0 +1,255 @@
+        checkout_status: str,
+        payment_id: str,
+    ) -> None:
+        if payment.status in PAYMENT_STATUS_FINAL:
+            # shouldn't happen or race-condition
+            return
+
+        status = get_payment_status(worldline_status, checkout_status=checkout_status)
+
+        # run this query as atomic update()
+        qs = SubmissionPayment.objects.filter(id=payment.id)
+        qs = qs.exclude(status__in=PAYMENT_STATUS_FINAL)
+        qs = qs.exclude(status=status)
+        res = qs.update(status=status, provider_payment_id=payment_id)
+
+        if res > 0:
+            payment.refresh_from_db()
+
+    def handle_return(
+        self,
+        request: Request,
+        payment: SubmissionPayment,
+        options: PaymentOptions,
+    ) -> HttpResponse:
+        with structlog.contextvars.bound_contextvars(
+            submission_uuid=str(payment.submission.uuid),
+            payment_uuid=str(payment.uuid),
+            plugin=self,
+            entrypoint="browser",
+        ):
+            log = logger.bind()
+            log.info("process_payment_status")
+
+            returnmac = payment.plugin_options.get("returnmac", "")
+            checkout_id = payment.plugin_options.get("checkout_id", "")
+
+            if (
+                not request.query_params.get("RETURNMAC") == returnmac
+                or not request.query_params.get("hostedCheckoutId") == checkout_id
+            ):
+                return HttpResponseBadRequest(b"Incorrect query parameters provided")
+
+            token = submission_status_token_generator.make_token(payment.submission)
+            status_url = request.build_absolute_uri(
+                reverse(
+                    "api:submission-status",
+                    kwargs={"uuid": payment.submission.uuid, "token": token},
+                )
+            )
+
+            client = options["merchant"].get_checkout_client()
+
+            try:
+                response = client.get_hosted_checkout(checkout_id)
+            except ApiException:
+                return HttpResponseServerError(b"Unable to retrieve checkout status")
+
+            payment_data = response.created_payment_output
+            checkout_status = response.status if response.status else ""
+            status = (
+                payment_data.payment.status
+                if payment_data and payment_data.payment and payment_data.payment.status
+                else ""
+            )
+
+            payment_provider_id = (
+                payment_data.payment.id
+                if payment_data and payment_data.payment and payment_data.payment.id
+                else ""
+            )
+
+            try:
+                self.apply_status(
+                    payment,
+                    status,
+                    checkout_status,
+                    payment_provider_id,
+                )
+            except KeyError:
+                return HttpResponseServerError(b"Unable to retrieve payment status")
```

Comment:
I think the `KeyError` is a leaky abstraction and it wasn't obvious were it's coming from. Perhaps a custom exception would make this more obvious:

```py
class StatusMappingError(Exception):
    pass
```

Reply:
This was also updated in https://github.com/open-formulieren/open-forms/commit/890e7b97eefdd0cd478114ab6613460587f1f5ca (see line 221-224).